python/fbprophet/diagnostics.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+import numpy as np
+import pandas as pd
+from functools import reduce
+from fbprophet import Prophet
+
+def _copy(model):
+    result = Prophet(
+        growth=model.growth,
+        n_changepoints=model.n_changepoints,
+        yearly_seasonality=model.yearly_seasonality,
+        weekly_seasonality=model.weekly_seasonality,
+        holidays=model.holidays,
+        seasonality_prior_scale=model.seasonality_prior_scale,
+        changepoint_prior_scale=model.changepoint_prior_scale,
+        holidays_prior_scale=model.holidays_prior_scale,
+        mcmc_samples=model.mcmc_samples,
+        interval_width=model.interval_width,
+        uncertainty_samples=model.uncertainty_samples
+    )
+    return result
+
+
+def _cutoff_points(te, horizon, period, k):
+    return [te - horizon - period * i for i in reversed(range(k))]
+
+
+def shf(model, horizon, k, period=None):
+    """Simulated Historical Forecasts.
+
+    Parameters
+    ----------
+    model: Prophet class object. Fitted Prophet model
+    horizon: string which has pd.Timedelta compatible style.
+        Forecast horizon
+    k: Int number. The number of forecasts point.
+    period: string which has pd.Timedelta compatible style or None, default None
+        Simulated Forecast will be done at every this period.
+        0.5 * horizon is used when it is None.
+
+    Returns
+    -------
+    A pd.DataFrame with the forecast, actual value and cutoff.
+    """
+    pd.to_timedelta
+    df = model.history.copy().reset_index(drop=True)
+    horizon = pd.Timedelta(horizon)
+    period = 0.5 * horizon if period is None else pd.Timedelta(period)
+    cutoffs = _cutoff_points(df['ds'].max(), horizon, period, k)
+
+    predicts = []
+    for cutoff in cutoffs:
+        # Generate new object with copying fitting options
+        m = _copy(model)
+        # Train model
+        m.fit(df[df['ds'] <= cutoff])
+        # Calculate yhat
+        index_predicted = (df['ds'] >= cutoff) & (df['ds'] <= cutoff + horizon)
```

Comment:
In the paper at the cutoff we were not using the cutoff date in the history (so strict inequality in line 71 here). Definitely either L71 or L73 should be strict, depending on if we want "cutoff" to mean data up to and including that date, or if we want it to mean up to but not including that date. I personally think it is more natural to have the cutoff date included in the training as is currently implemented in L71 (and so not in the validation). We are about to update the paper and I can make sure it is consistent there. If you think it is more natural to have the cutoff date not included in the history (and so in the validation) I'm open to that also.